conans/model/build_info.py

```diff
@@ -111,10 +191,75 @@ def __init__(self, root_folder):
         self.libdirs.append(DEFAULT_LIB)
         self.bindirs.append(DEFAULT_BIN)
         self.resdirs.append(DEFAULT_RES)
-        self.builddirs.append("")
+        self.builddirs.append(DEFAULT_BUILD)
         # public_deps is needed to accumulate list of deps for cmake targets
         self.public_deps = []
         self.configs = {}
+        self._deps = OrderedDict()
+
+    def _check_dirs_values(self):
+        default_dirs_mapping = {
+            "includedirs": [DEFAULT_INCLUDE],
+            "libdirs": [DEFAULT_LIB],
+            "bindirs": [DEFAULT_BIN],
+            "resdirs": [DEFAULT_RES],
+            "builddirs": [DEFAULT_BUILD],
+            "srcdirs": []
+        }
+        msg_template = "Using Components and global '{}' values ('{}') is not supported"
+        for dir_name in ["includedirs", "libdirs", "bindirs", "builddirs"]:
+            dirs_value = getattr(self, dir_name)
+            if dirs_value is not None and dirs_value != default_dirs_mapping[dir_name]:
+                raise ConanException(msg_template.format(dir_name, dirs_value))
+
+    def _clear_dirs_values(self):
+        default_dirs_mapping = {
+            "includedirs": [DEFAULT_INCLUDE],
+            "libdirs": [DEFAULT_LIB],
+            "bindirs": [DEFAULT_BIN],
+            "resdirs": [DEFAULT_RES],
+            "builddirs": [DEFAULT_BUILD],
+            "srcdirs": []
+        }
+        for dir_name in ["includedirs", "libdirs", "bindirs", "builddirs"]:
+            if getattr(self, dir_name) == default_dirs_mapping[dir_name]:
+                self.__dict__[dir_name] = None
+
+    @property
+    def libs(self):
+        if self._deps:
+            deps = [v for v in self._deps.values()]
+            deps_sorted = sorted(deps, key=lambda component: len(component.deps))
+            result = []
+            for dep in deps_sorted:
+                for sys_dep in dep.system_deps:
+                    if sys_dep and sys_dep not in result:
+                        result.append(sys_dep)
+                if dep.lib:
+                    result.append(dep.lib)
+            return result
+        else:
+            return self._libs
+
+    @libs.setter
+    def libs(self, libs):
+        if self._deps:
+            raise ConanException("Setting first level libs is not supported when Components are "
+                                 "already in use")
+        self._libs = libs
+
+    def __getitem__(self, key):
+        if self._libs or self._exes:
+            raise ConanException("Usage of Components with '.libs' or '.exes' values is not allowed")
+        self._clear_dirs_values()
+        self._check_dirs_values()
+        if key not in self._deps.keys():
+            self._deps[key] = Component(key, self.rootpath)
+        return self._deps[key]
+
+    @property
+    def deps(self):
```

Comment:
I think it makes more sense to call this ``components`` instead, as it might be confusing with ``self.deps_cpp_info[""].deps`` which points to nodes in the graph. Whereas this ``deps`` here points to components/dependencies inside the same node. WDYT?